### ISSUE TITLE
Suggestion: Components Panel

### ISSUE THREAD

**Issue by SolarLune** [GoranM/bdx#613]
Hey, there. 

I've been thinking about this for awhile, and I think a Components panel might be cool to implement. The general idea behind it is that it would allow you to add, remove, and tweak components that you create beforehand in code. This way, you don't have to add a class for a GameObject just to add components. Basically, it would make the "component-only" workflow a bit more clean and editable from Blender.

The general idea is that the BDX add-on would search through the source files for lines like:

```
class xxxxxxxxx extends Component
```

and present them as options when you go to add a Component to a GameObject. You'd also be able to get and set a component's properties from Blender. The hard part would be the implementation, but I think the goal would be to allow you to, say, add a value name and type to a Component's source code, and then have that show up in Blender as something you can edit and change. When you press P to play the game, it creates the component, alters it as necessary to match what was in Blender, and adds it to the GameObject.

Thoughts?

**Comment by rafcolson** [comment]
I like the idea. I think it could work like the "Add Modifier" operator. Instead it would be an "Add Component" operator in the BDX Object Panel, beneath the Game Properties panel.

Public fields could then be the properties being shown in the Component Panel. I think it will add some limitations to the way a Component is coded though. For example, the constructor cannot contain arguments other than the GameObject itself, unless the Component Panel not only allows to specify fields, but also constructor arguments. But that would become a little confusing.

**Comment by GoranM** [comment]
> This way, you don't have to add a class for a GameObject just to add components.

You don't have to do that now; As long as the component in question only requires/manipulates methods/fields defined on GameObject, it can be added to any GameObject:

```java
// add Explosion component to every "Bomb" instance:
for (GameObject g : scene.objects){
    if (g.name().equals("Bomb")){
        g.components.add(new Explosion(g));
    }
}
```

> The general idea is that the BDX add-on would search through the source files

It would be simpler to establish a convention where components you want to have available reside in the `com.organization.project.components` package.

Although, even then, in order to get something like this to work, you would need to do some code generation, like we do for scene instantiators, because the HTML5 backend doesn't support the kind of reflection that would allow us to instantiate arbitrary classes dynamically.

**Comment by SolarLune** [comment]
> You don't have to do that now; As long as the component in question only requires/manipulates methods/fields defined on GameObject, it can be added to any GameObject:

Yes, but it's awkward to run code like that on every frame just to make sure that added objects that don't have a specific component get one. It quickly spirals out of control for larger and larger scenes and more and more objects.

> It would be simpler to establish a convention where components you want to have available reside in the com.organization.project.components package.

That would be convenient, but considering that that would force Components to exist under a specific directory and namespace, it's probably not a wise solution.

> Although, even then, in order to get something like this to work, you would need to do some code generation, like we do for scene instantiators, because the HTML5 backend doesn't support the kind of reflection that would allow us to instantiate arbitrary classes dynamically.

Hmm, we'll have to think about how best to approach this, then.

**Comment by GoranM** [comment]
> Hmm, we'll have to think about how best to approach this, then.

I'm wondering if we really need this.

The complexity increase is significant, in both the plug-in, and the engine, but simply writing a class doesn't seem difficult enough to justify that increase.

**Comment by SolarLune** [comment]
> The complexity increase is significant, in both the plug-in, and the engine, but simply writing a class doesn't seem difficult enough to justify that increase.

It is true that it would be a bit more complex for a relatively little gain. Maybe there's another way around it that doesn't increase complexity so much, but still adds the core functionality to register Components without needing to create classes for GameObjects?

What about, say, adding something to that allows one to register GameObjects' names with sets of components? Something like this run before anything else in the game scene:

```java

import com.game.components.InputMovement;
import com.game.components.Inventory;

public void init(){
    GameObject.register("Player", InputMovement.class, Inventory.class);
}

```

Then any time a GameObject by the name of "Player" is added, it will have the registered components instantiated and added to it?

**Comment by GoranM** [comment]
I don't really understand how this is simpler.

You still have to write code (be it manually, or via some generator) to specify the binding between a particular object and a set of components.

The fact that you're using some register function instead of a class doesn't change that.

**Comment by SolarLune** [comment]
That's true, but now I don't have to have multiple files for multiple classes. Now I can have one file that I probably would have anyway either way (a class that runs before anything else to register components, set up the game state, bind inputs, etc), and just the components necessary for each type of GameObject. 

It's kind of a waste to have, say, the Player's Components and a Player class, the different Components for the Enemies and a class for each Enemy, etc. This suggestion would be in line with a truly component based game design. 

A weakness I do see is the inability to customize the components added - maybe there's another way to do this that would provide for that.

**Comment by GoranM** [comment]
> It's kind of a waste to have, say, the Player's Components and a Player class, the different Components for the Enemies and a class for each Enemy, etc.

I don't see how it's a waste; It's an isolated point of specification, where you can set components, define custom methods, and control initialization. It's a proven encapsulation vehicle, and we can use it without having to introduce additional system complexity (and the additional problems that would bring).

**Comment by SolarLune** [comment]
Well, I suppose so. I guess it's just more trouble than it's worth. I'll close this, then.

**Comment by SolarLune** [comment]
Re-opening this because I still feel like this might be a good idea. Seems like with hot-reloading implemented (how we can export a scene from Blender for BDX to auto reload), it's possible to cut down on the need to restart your game if you use components added and tweaked via Blender. 

[I made progress](https://github.com/SolarLune/bdx/commit/da603a555f2e01b3b21a9d31e332f295196df831) on actually implementing this fully - only thing missing is messing with Components' properties. What do you think?

**Comment by GoranM** [comment]
I think my original arguments still hold.

**Comment by SolarLune** [comment]
Your original argument about the code cost being too great?

I feel like it is somewhat more code, but that amount of code is pretty "light" in terms of "depth" (it's not like it's a bunch of hard calculations or particularly difficult to maintain or understand code). The code is basically necessary for the concept at hand, and most of it is simple UI / panel stuff. 

If you think it might be too heavy all the same, then maybe there's a simpler, more generic alternative. For example, maybe we could just have the Component panel. It wouldn't have auto-detection, it would just hold the name of the Component and allow you to define properties that are tied to that particular Component. It would store the name of the Component and its properties in a map for the user to work with, something like:

```java

public void init() {

    for (JsonValue comp : props.get("components")) {

        if (comp.name.equals("Halo"))
            components.add(new Halo(this));
        else if (comp.name.equals("Gauge"))
            components.add(new Gauge(this, comp.get("value").asFloat()));

    }

}

```

Or, maybe we could define a new property class that replaces the current Game Properties section and expands on it with more useful types (like Components, Vectors, Colors, GameObject selectors (which would just convert into ordinary Strings), etc).

**Comment by SolarLune** [comment]
Oh, or we could add a callback in Scene that gets called when a GameObject is added. 

That way, the user could define it to add components to GameObjects when necessary. For example, it could add a "HealthBar" component to GameObjects that have a "health" property when they're created.

This'd be cheap and still be usable and extendable to easily have a sort of game-wide GUI for component systems.

**Comment by GoranM** [comment]
> maybe we could just have the Component panel. It wouldn't have auto-detection, it would just hold the name of the Component and allow you to define properties that are tied to that particular Component.

That sounds better.

> Or, maybe we could define a new property class that replaces the current Game Properties section and expands on it with more useful types (like Components, Vectors, Colors, GameObject selectors (which would just convert into ordinary Strings), etc).

I don't think there's a need for that. Being able to define components, and their properties, seems sufficient for the core benefits.

Also, I'm not sure if there's a need for any kind of "callback": Given a generated instantiator, you should just be able to call it when the game object is created.

**Comment by SolarLune** [comment]
OK, I made more progress on this. I forgot that because the Instantiator is compiled, changing and exporting the components in the blend file won't influence an already-running game, so that benefit is lost, unfortunately (which is a major strength, in my opinion).

The options I mentioned were alternatives. With a callback, we wouldn't really need a panel at all to get roughly the same functionality, since the user could define it to create specific components for specific properties for each GameObject spawned.

**Comment by GoranM** [comment]
> I forgot that because the Instantiator is compiled, changing and exporting the components in the blend file won't influence an already-running game, so that benefit is lost

Unless you're using components that are not referenced in the existing instantiator file, there should be no need to recompile.

**Comment by SolarLune** [comment]
Hmm, that's not the results that I'm seeing. 

What's your idea for this? Because [my branch](https://github.com/SolarLune/bdx/commit/201f519be3b6baedd28956ca59c4a7fd717e33a5) reads the Blender data for the object and then edits an Instantiator to create and add the relevant components to the GameObject after creation. Should it work some other way?

**Comment by GoranM** [comment]
> Or were you thinking that it should be what I mentioned earlier - that the component data should be exported as part of the JSON for each GameObject to read in manually?

I'm not really sure what you mean by "manually", but yes, I think it would be ok to export component data as JSON.
